pkg/objstore/oss/oss.go

```diff
@@ -0,0 +1,380 @@
+package oss
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"math"
+	"math/rand"
+	"net/http"
+	"os"
+	"strconv"
+	"strings"
+	"testing"
+	"time"
+
+	alioss "github.com/aliyun/aliyun-oss-go-sdk/oss"
+
+	"github.com/go-kit/kit/log"
+	"github.com/pkg/errors"
+	"github.com/thanos-io/thanos/pkg/objstore"
+	"github.com/thanos-io/thanos/pkg/runutil"
+	"gopkg.in/yaml.v2"
+)
+
+type Config struct {
+	Endpoint        string `yaml:"endpoint"`
+	Bucket          string `yaml:"bucket"`
+	AccessKeyID     string `yaml:"access_key_id"`
+	AccessKeySecret string `yaml:"access_key_secret"`
+}
+
+type Bucket struct {
+	name   string
+	logger log.Logger
+	client *alioss.Client
+	config Config
+	bucket *alioss.Bucket
+}
+
+func NewTestBucket(t testing.TB) (objstore.Bucket, func(), error) {
+	c := configFromEnv()
+	err := func(conf Config) error {
+		if conf.Endpoint == "" {
+			return errors.New("no aliyun oss endpoint in config file")
+		}
+		if conf.AccessKeyID == "" {
+			return errors.New("access_key_id is not present in config file")
+		}
+		if conf.AccessKeySecret == "" {
+			return errors.New("access_key_secret is not present in config file")
+		}
+		return nil
+	}(c)
+	if err != nil {
+		return nil, nil, err
+	}
+	if c.Bucket != "" && os.Getenv("THANOS_ALLOW_EXISTING_BUCKET_USE") == "true" {
+		t.Log("ALIYUNOSS_BUCKET is defined. Normally this tests will create temporary bucket " +
+			"and delete it after test. Unset ALIYUNOSS_BUCKET env variable to use default logic. If you really want to run " +
+			"tests against provided (NOT USED!) bucket, set THANOS_ALLOW_EXISTING_BUCKET_USE=true.")
+		return NewTestBucketFromConfig(t, c, true)
+	}
+	return NewTestBucketFromConfig(t, c, false)
+}
+
+func calculateChunks(name string, r io.Reader) (int, int64, error) {
```

Comment:
I think we shouldn't do this type castin, let's just take an interface which has `io.Reader` & `Size()` methods instead of supporting only `os.File` & `strings.Reader` types.

Reply:
I am a little ambiguous about this. We take S3 as an example, could you please give me comments in more details. Thanks!

Reply:
I think we can do this in another PR. As this would require quite a lot of refactoring.

If we change the interface in https://github.com/thanos-io/thanos/blob/master/pkg/objstore/objstore.go#L26
 from:
```
Upload(ctx context.Context, name string, r io.Reader) error
```

to

```
type ReadSizer interface{
   io.Reader
   Size() int64
} 

Upload(ctx context.Context, name string, r ReadSizer) error
```

We would avoid this hack entirely.

@bwplotka what do you think?

Reply:
I guess there is no way to have this working without knowing the size beforehand? :thinking:

Reply:
We could change interface, but the S3 example is bad - we should never rely on size as it block streaming case. However the truth is there is no case we need streaming currently for write (:

Reply:
Yea ok it looks like for multipart need to do on our own: https://github.com/aliyun/aliyun-oss-go-sdk/blob/d0a2d03230103776946b757cbc05aeecb523d9c9/oss/multipart_test.go#L180

Reply:
The other option is to actually read all the bytes into memory and then we would know the size, which sucks for memory requirements. 

I would prefer to have `ReadSizer`, it actually helps upload case, as the uploader client can split the object into chunks, make it parallel, etc.

Reply:
Well that's the thing, for streaming purposes you buffer let's say 128KB then if you buffer more you know you need multipart. While you traverse the file you stream parts until end of the reader (: I think that's what GCS is doing for example. Essentially you can implement this without knowing size beforehand.

Reply:
In fact this will be required if we would like to have this: https://github.com/thanos-io/thanos/issues/1673

Reply:
Yeah you are right. Actually in this case we shouldn't need a Size() method, we can just do what you said and upload parts the way you described.